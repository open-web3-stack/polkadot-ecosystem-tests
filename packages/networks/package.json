{
	"name": "@e2e-test/networks",
	"type": "module",
	"dependencies": {
		"@acala-network/chopsticks-testing": "^0.13.1",
		"@e2e-test/shared": "workspace:*"
	},
	"main": "./dist/cjs/index.js",
	"module": "./dist/esm/index.js",
	"files": [
		"*",
		"!tsconfig.json",
		"!**/*.test.ts"
	],
	"exports": {
		".": {
			"require": "./dist/cjs/index.js",
			"import": "./dist/esm/index.js",
			"default": "./dist/esm/index.js"
		},
		"./chains": {
			"require": "./dist/cjs/chains/index.js",
			"import": "./dist/esm/chains/index.js",
			"default": "./dist/esm/chains/index.js"
		},
		"./*": {
			"require": "./dist/cjs/*.js",
			"import": "./dist/esm/*.js",
			"default": "./dist/esm/*.js"
		}
	}
}
