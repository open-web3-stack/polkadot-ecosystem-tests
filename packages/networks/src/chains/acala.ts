import { defaultAccount } from '@e2e-test/shared'

import { defineChain } from '../defineChain.js'

const custom = {
  acala: {
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
    stableToken: 'AUSD',

    paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
    dot: { Token: 'DOT' },
    ldot: { Token: 'LDOT' },
    dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
    wbtc: { ForeignAsset: 5 },
    ausd: { Token: 'AUSD' },
    aca: { Token: 'ACA' },
    lcdot: { LiquidCrowdloan: 13 },
  },
  karura: {
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
    stableToken: 'KUSD',

    paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
    ksm: { Token: 'KSM' },
    lksm: { Token: 'LKSM' },
    usdt: { ForeignAsset: 7 },
    rmrk: { ForeignAsset: 0 },
    dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
    ausd: { Token: 'KUSD' },
    kar: { Token: 'KAR' },
  },
}

const getInitStorages = (config: typeof custom.acala | typeof custom.karura) => ({
  System: {
    account: [[[defaultAccount.alice.address], { providers: 4, data: { free: 10 * 1e12 } }]],
  },
  Tokens: {
    accounts: [
      [[defaultAccount.alice.address, { Token: config.relayToken }], { free: 10 * 1e12 }],
      [[defaultAccount.alice.address, { Token: config.relayLiquidToken }], { free: 100 * 1e12 }],
      [[defaultAccount.alice.address, { Token: config.stableToken }], { free: 1000 * 1e12 }],
    ],
  },
  Sudo: {
    key: defaultAccount.alice.address,
  },
  EvmAccounts: {
    accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], defaultAccount.alice.address]],
    evmAddresses: [[[defaultAccount.alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
  },
  Homa: {
    // avoid impact test outcome
    $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
    // so that bump era won't trigger unbond
    relayChainCurrentEra: 100,
  },
  PolkadotXcm: {
    // avoid sending xcm version change notifications to makes things faster
    $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
  },
})

export const acala = defineChain({
  name: 'acala',
  endpoint: 'wss://acala-rpc.aca-api.network',
  paraId: 2000,
  custom: custom.acala,
  initStorages: getInitStorages(custom.acala),
})

export const karura = defineChain({
  name: 'karura',
  endpoint: 'wss://karura-rpc.aca-api.network',
  paraId: 2000,
  custom: custom.karura,
  initStorages: getInitStorages(custom.karura),
})
