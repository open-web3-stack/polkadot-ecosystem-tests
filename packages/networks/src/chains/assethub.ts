import { defaultAccounts } from '../defaultAccounts.js'
import { defineChain } from '../defineChain.js'

const custom = {
  assetHubPolkadot: {
    dot: { Concrete: { parents: 1, interior: 'Here' } },
    usdt: { Concrete: { parents: 0, interior: { X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
    usdtIndex: 1984,
  },
  assetHubKusama: {
    ksm: { Concrete: { parents: 1, interior: 'Here' } },
    usdt: { Concrete: { parents: 0, interior: { X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
    usdtIndex: 1984,
  },
}

const getInitStorages = (config: typeof custom.assetHubPolkadot | typeof custom.assetHubKusama) => ({
  System: {
    account: [[[defaultAccounts.alice.address], { providers: 1, data: { free: 1000e10 } }]],
  },
  Assets: {
    account: [
      [[config.usdtIndex, defaultAccounts.alice.address], { balance: 1000e6 }], // USDT
    ],
  },
})

export const assetHubPolkadot = defineChain({
  name: 'assetHubPolkadot',
  endpoint: 'wss://rpc-asset-hub-polkadot.luckyfriday.io',
  paraId: 1000,
  custom: custom.assetHubPolkadot,
  initStorages: getInitStorages(custom.assetHubPolkadot),
})

export const assetHubKusama = defineChain({
  name: 'assetHubKusama',
  endpoint: 'wss://rpc-asset-hub-kusama.luckyfriday.io',
  paraId: 1000,
  custom: custom.assetHubKusama,
  initStorages: getInitStorages(custom.assetHubKusama),
})
