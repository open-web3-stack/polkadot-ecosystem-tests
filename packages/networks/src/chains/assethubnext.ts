import { defaultAccounts, defaultAccountsSr25519 } from '../defaultAccounts.js'
import { defineChain } from '../defineChain.js'

const custom = {
  assetHubNextWestend: {
    wnd: { Concrete: { parents: 1, interior: 'Here' } },
    usdt: { Concrete: { parents: 0, interior: { X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
    usdtIndex: 1984,
  },
}

const getInitStorages = (config: typeof custom.assetHubNextWestend) => ({
  System: {
    account: [
      [[defaultAccounts.alice.address], { providers: 1, data: { free: 1000e10 } }],
      [[defaultAccountsSr25519.alice.address], { providers: 1, data: { free: 1000e10 } }],
    ],
  },
  Assets: {
    account: [
      [[config.usdtIndex, defaultAccounts.alice.address], { balance: 1000e6 }], // USDT
    ],
  },
})

export const assetHubNextWestend = defineChain({
  name: 'assetHubNextWestend',
  endpoint: 'wss://westend-asset-hub-next-rpc.parity-chains-scw.parity.io',
  paraId: 1100,
  custom: custom.assetHubNextWestend,
  initStorages: getInitStorages(custom.assetHubNextWestend),
})
