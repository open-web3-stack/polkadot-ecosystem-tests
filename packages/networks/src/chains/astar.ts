import { defaultAccounts } from '../defaultAccounts.js'
import { defineChain } from '../defineChain.js'

const custom = {
  astar: {
    relayToken: '340282366920938463463374607431768211455',
    aUSDToken: '18446744073709551617',

    paraAccount: '13YMK2eZzuFY1WZGagpYtTgbWBWGdoUD2CtrPj1mQPjY8Ldc',
    dot: 340282366920938463463374607431768211455n,
    astr: { Concrete: { parents: 0, interior: 'Here' } },
    aca: 18446744073709551616n,
    usdt: 4294969280n,
  },
  shiden: {
    relayToken: '340282366920938463463374607431768211455',
    aUSDToken: '18446744073709551616',

    paraAccount: 'F7fq1jNy74AqkJ1DP4KqSrWtnTGtXfNVoDwFhTvvPxUvJaq',
    ksm: 340282366920938463463374607431768211455n,
    sdn: { Concrete: { parents: 0, interior: 'Here' } },
    kar: 18446744073709551618n,
    usdt: 4294969280n,
  },
}

const getInitStorages = (config: typeof custom.astar | typeof custom.shiden) => ({
  System: {
    account: [[[defaultAccounts.alice.address], { providers: 1, data: { free: '100000000000000000000' } }]],
  },
  Assets: {
    account: [
      [[config.relayToken, defaultAccounts.alice.address], { balance: 10 * 1e12 }],
      [[config.aUSDToken, defaultAccounts.alice.address], { balance: 10 * 1e12 }],
    ],
  },
  Sudo: {
    key: defaultAccounts.alice.address,
  },
  PolkadotXcm: {
    // avoid sending xcm version change notifications to makes things faster
    $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
  },
})

export const astar = defineChain({
  name: 'astar',
  paraId: 2006,
  endpoint: 'wss://astar-rpc.dwellir.com',
  custom: custom.astar,
  initStorages: getInitStorages(custom.astar),
})

export const shiden = defineChain({
  name: 'shiden',
  paraId: 2007,
  endpoint: 'wss://shiden-rpc.dwellir.com',
  custom: custom.shiden,
  initStorages: getInitStorages(custom.shiden),
})
