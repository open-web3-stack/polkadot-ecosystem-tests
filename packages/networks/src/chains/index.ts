export * from './acala.js'
export * from './assethub.js'
export * from './astar.js'
export * from './hydraDX.js'
export * from './moonbeam.js'
export * from './polkadot.js'
