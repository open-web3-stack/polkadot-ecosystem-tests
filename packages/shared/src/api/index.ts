import type { ApiPromise } from '@polkadot/api'

export type TransferType = 'Teleport' | 'LocalReserve' | 'DestinationReserve' | 'RemoteReserve'

export const xtokens = {
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV4: (acc: any) => ({
    V4: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainAccountId20V3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV4: (paraId: number) => (acc: any) => ({
    V4: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited',
      ),
}

export const xcmPallet = {
  relaychainV4: {
    V4: {
      parents: 1,
      interior: 'Here',
    },
  },
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  parachainV4: (parents: number, paraId: any) => ({
    V4: {
      parents,
      interior: {
        X1: [{ Parachain: paraId }],
      },
    },
  }),
  limitedTeleportAssets:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedTeleportAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  // network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited',
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited',
      ),
  teleportAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).teleportAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
      ),
  executeXCM:
    (xcm: any, max_weight: any) =>
    ({ api }: { api: ApiPromise }) => {
      return (api.tx.xcmPallet || api.tx.polkadotXcm).execute(xcm, max_weight)
    },
  transferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).transferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited',
      ),
  transferAssetsUsingType:
    (dest: any, tokens: any[], assetTransferType: TransferType, remoteFeesId: any, feesTransferType: TransferType) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).transferAssetsUsingTypeAndThen(
        dest,
        { V4: tokens },
        assetTransferType,
        { V4: remoteFeesId },
        feesTransferType,
        {
          V4: [
            {
              DepositAsset: {
                assets: {
                  Wild: {
                    AllCounted: 2,
                  },
                },
                beneficiary: {
                  parents: 0,
                  interior: { X1: [{ AccountId32: { id: acc } }] },
                },
              },
            },
          ],
        },
        'Unlimited',
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  totalIssuance: ({ api }: { api: ApiPromise }) => api.query.balances.totalIssuance(),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  foreignAssets:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.foreignAssets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
