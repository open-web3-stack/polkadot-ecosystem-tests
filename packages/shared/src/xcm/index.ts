export * from './runXcmPalletDown.js'
export * from './runXcmPalletHorizontal.js'
export * from './runXcmPalletUp.js'
export * from './runXtokenstHorizontal.js'
export * from './runXtokensUp.js'
