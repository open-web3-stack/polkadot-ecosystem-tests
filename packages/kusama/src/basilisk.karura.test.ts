import { afterAll, beforeEach, describe } from 'vitest'

import { basilisk, karura, kusama } from '@e2e-test/networks/chains'
import { captureSnapshot, createNetworks } from '@e2e-test/networks'
import { defaultAccount } from '@e2e-test/shared'
import { query, tx } from '@e2e-test/shared/api'
import { runXtokenstHorizontal } from '@e2e-test/shared/xcm'

describe('basilisk & karura', async () => {
  const [karuraClient, basiliskClient, kusamaClient] = await createNetworks(karura, basilisk, kusama)

  const restoreSnapshot = captureSnapshot(karuraClient, basiliskClient, kusamaClient)

  beforeEach(restoreSnapshot)

  afterAll(async () => {
    await karuraClient.teardown()
    await basiliskClient.teardown()
    await kusamaClient.teardown()
  })

  runXtokenstHorizontal('karura transfer KSM to basilisk', async () => {
    return {
      fromChain: karuraClient,
      toChain: basiliskClient,
      routeChain: kusamaClient,
      isCheckUmp: true,
      toAccount: defaultAccount.bob,
      fromBalance: query.tokens(karura.custom.ksm),
      toBalance: query.tokens(basilisk.custom.relayToken),
      tx: tx.xtokens.transfer(karura.custom.ksm, 10n ** 12n, tx.xtokens.parachainV4(basilisk.paraId!)),
    }
  })

  runXtokenstHorizontal('basilisk transfer KSM to karura', async () => {
    return {
      fromChain: basiliskClient,
      toChain: karuraClient,
      routeChain: kusamaClient,
      isCheckUmp: true,
      toAccount: defaultAccount.bob,
      fromBalance: query.tokens(basilisk.custom.relayToken),
      toBalance: query.tokens(karura.custom.ksm),
      tx: tx.xtokens.transfer(basilisk.custom.relayToken, 10n ** 12n, tx.xtokens.parachainV4(karura.paraId!)),
    }
  })

  runXtokenstHorizontal('basilisk transfer BSX to karura', async () => {
    return {
      fromChain: basiliskClient,
      toChain: karuraClient,
      fromBalance: query.balances,
      toBalance: query.tokens(karura.custom.bsx),
      tx: tx.xtokens.transfer(basilisk.custom.bsx, 10n ** 15n, tx.xtokens.parachainV4(karura.paraId!)),
    }
  })

  runXtokenstHorizontal('karura transfer BSX to basilisk', async () => {
    await karuraClient.dev.setStorage({
      Tokens: {
        Accounts: [[[defaultAccount.alice.address, karura.custom.bsx], { free: 10n * 10n ** 15n }]],
      },
    })

    return {
      fromChain: karuraClient,
      toChain: basiliskClient,
      fromBalance: query.tokens(karura.custom.bsx),
      toBalance: query.balances,
      tx: tx.xtokens.transfer(karura.custom.bsx, 10n ** 15n, tx.xtokens.parachainV4(basilisk.paraId!)),
    }
  })
})
