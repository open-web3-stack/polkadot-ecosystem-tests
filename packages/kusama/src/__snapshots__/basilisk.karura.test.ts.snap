// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basilisk & karura > basilisk transfer BSX to karura > balance on from chain 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": "(rounded 999000000000000000)",
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 1,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`basilisk & karura > basilisk transfer BSX to karura > balance on to chain 1`] = `
{
  "free": "(rounded 1000000000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`basilisk & karura > basilisk transfer BSX to karura > from chain hrmp messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "reserveAssetDeposited": [
              {
                "fun": {
                  "fungible": 1000000000000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "parachain": 2090,
                        },
                        {
                          "generalIndex": 0,
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 1000000000000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "parachain": 2090,
                        },
                        {
                          "generalIndex": 0,
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0xd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 1,
            },
          },
        ],
      },
    ],
    "recipient": 2000,
  },
]
`;

exports[`basilisk & karura > basilisk transfer BSX to karura > to chain xcm events 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Sibling": "(rounded 2100)",
      },
      "success": true,
      "weightUsed": {
        "proofSize": 0,
        "refTime": 800000000,
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`basilisk & karura > basilisk transfer BSX to karura > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000000,
          },
          "id": {
            "interior": {
              "X2": [
                {
                  "Parachain": "(rounded 2100)",
                },
                {
                  "GeneralIndex": 0,
                },
              ],
            },
            "parents": 1,
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": 2000,
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000000,
        },
        "id": {
          "interior": {
            "X2": [
              {
                "Parachain": "(rounded 2100)",
              },
              {
                "GeneralIndex": 0,
              },
            ],
          },
          "parents": 1,
        },
      },
      "sender": "bXjgry3bkMHToRvMRedUhtPzEUUXdhyVX2UwY8hY156tWH4YB",
    },
    "method": "TransferredAssets",
    "section": "xTokens",
  },
]
`;

exports[`basilisk & karura > basilisk transfer KSM to karura > balance on from chain 1`] = `
{
  "free": 999000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`basilisk & karura > basilisk transfer KSM to karura > balance on to chain 1`] = `
{
  "free": "(rounded 998000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`basilisk & karura > basilisk transfer KSM to karura > from chain ump messages 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 500000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "depositReserveAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "dest": {
            "interior": {
              "x1": {
                "parachain": 2000,
              },
            },
            "parents": 0,
          },
          "maxAssets": 1,
          "xcm": [
            {
              "buyExecution": {
                "fees": {
                  "fun": {
                    "fungible": 500000000000,
                  },
                  "id": {
                    "concrete": {
                      "interior": {
                        "here": null,
                      },
                      "parents": 1,
                    },
                  },
                },
                "weightLimit": {
                  "unlimited": null,
                },
              },
            },
            {
              "depositAsset": {
                "assets": {
                  "wild": {
                    "all": null,
                  },
                },
                "beneficiary": {
                  "interior": {
                    "x1": {
                      "accountId32": {
                        "id": "0xd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
                        "network": {
                          "any": null,
                        },
                      },
                    },
                  },
                  "parents": 0,
                },
                "maxAssets": 1,
              },
            },
          ],
        },
      },
    ],
  },
]
`;

exports[`basilisk & karura > basilisk transfer KSM to karura > route chain xcm events 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Ump": {
          "Para": "(rounded 2100)",
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 7200)",
        "refTime": "(rounded 640000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`basilisk & karura > basilisk transfer KSM to karura > to chain xcm events 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": "Parent",
      "success": true,
      "weightUsed": {
        "proofSize": 0,
        "refTime": 1000000000,
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`basilisk & karura > basilisk transfer KSM to karura > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000,
          },
          "id": {
            "interior": "Here",
            "parents": 1,
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": 2000,
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000,
        },
        "id": {
          "interior": "Here",
          "parents": 1,
        },
      },
      "sender": "bXjgry3bkMHToRvMRedUhtPzEUUXdhyVX2UwY8hY156tWH4YB",
    },
    "method": "TransferredAssets",
    "section": "xTokens",
  },
]
`;

exports[`basilisk & karura > karura transfer BSX to basilisk > balance on from chain 1`] = `
{
  "free": 9000000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`basilisk & karura > karura transfer BSX to basilisk > balance on to chain 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": 978000000000000,
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`basilisk & karura > karura transfer BSX to basilisk > from chain hrmp messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "withdrawAsset": [
              {
                "fun": {
                  "fungible": 1000000000000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x1": {
                        "generalIndex": 0,
                      },
                    },
                    "parents": 0,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 1000000000000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x1": {
                        "generalIndex": 0,
                      },
                    },
                    "parents": 0,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0xd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 1,
            },
          },
        ],
      },
    ],
    "recipient": 2090,
  },
]
`;

exports[`basilisk & karura > karura transfer BSX to basilisk > to chain xcm events 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Sibling": 2000,
      },
      "success": true,
      "weightUsed": {
        "proofSize": 0,
        "refTime": 400000000,
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`basilisk & karura > karura transfer BSX to basilisk > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000000,
          },
          "id": {
            "interior": {
              "X2": [
                {
                  "Parachain": "(rounded 2100)",
                },
                {
                  "GeneralIndex": 0,
                },
              ],
            },
            "parents": 1,
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(rounded 2100)",
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000000,
        },
        "id": {
          "interior": {
            "X2": [
              {
                "Parachain": "(rounded 2100)",
              },
              {
                "GeneralIndex": 0,
              },
            ],
          },
          "parents": 1,
        },
      },
      "sender": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "TransferredAssets",
    "section": "xTokens",
  },
]
`;

exports[`basilisk & karura > karura transfer KSM to basilisk > balance on from chain 1`] = `
{
  "free": 9000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`basilisk & karura > karura transfer KSM to basilisk > balance on to chain 1`] = `
{
  "free": "(rounded 998000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`basilisk & karura > karura transfer KSM to basilisk > from chain ump messages 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 500000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "depositReserveAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "dest": {
            "interior": {
              "x1": {
                "parachain": 2090,
              },
            },
            "parents": 0,
          },
          "maxAssets": 1,
          "xcm": [
            {
              "buyExecution": {
                "fees": {
                  "fun": {
                    "fungible": 500000000000,
                  },
                  "id": {
                    "concrete": {
                      "interior": {
                        "here": null,
                      },
                      "parents": 1,
                    },
                  },
                },
                "weightLimit": {
                  "unlimited": null,
                },
              },
            },
            {
              "depositAsset": {
                "assets": {
                  "wild": {
                    "all": null,
                  },
                },
                "beneficiary": {
                  "interior": {
                    "x1": {
                      "accountId32": {
                        "id": "0xd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
                        "network": {
                          "any": null,
                        },
                      },
                    },
                  },
                  "parents": 0,
                },
                "maxAssets": 1,
              },
            },
          ],
        },
      },
    ],
  },
]
`;

exports[`basilisk & karura > karura transfer KSM to basilisk > route chain xcm events 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Ump": {
          "Para": 2000,
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 7200)",
        "refTime": "(rounded 640000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`basilisk & karura > karura transfer KSM to basilisk > to chain xcm events 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": "Parent",
      "success": true,
      "weightUsed": {
        "proofSize": 0,
        "refTime": 500000000,
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`basilisk & karura > karura transfer KSM to basilisk > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000,
          },
          "id": {
            "interior": "Here",
            "parents": 1,
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(rounded 2100)",
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000,
        },
        "id": {
          "interior": "Here",
          "parents": 1,
        },
      },
      "sender": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "TransferredAssets",
    "section": "xTokens",
  },
]
`;
